src/main/scala/viper/gobra/frontend/info/implementation/typing/ProgramTyping.scala

```diff
@@ -0,0 +1,95 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2020 ETH Zurich.
+
+package viper.gobra.frontend.info.implementation.typing
+
+import org.bitbucket.inkytonik.kiama.util.Entity
+import org.bitbucket.inkytonik.kiama.util.Messaging.{Messages, error}
+import viper.gobra.ast.frontend.{PExpression, POld, PPackage, PProgram, PVarDecl}
+import viper.gobra.frontend.info.base.{SymbolTable => st}
+import viper.gobra.frontend.info.implementation.TypeInfoImpl
+import viper.gobra.frontend.info.implementation.property.{AssignMode, StrictAssignMode}
+import viper.gobra.util.Violation
+
+trait ProgramTyping extends BaseTyping { this: TypeInfoImpl =>
+
+  lazy val wellDefProgram: WellDefinedness[PProgram] = createWellDef {
+    case PProgram(_, posts, imports, members) =>
+      // Obtains global variable declarations sorted by the order in which they appear in the file
+      val sortedByPosDecls: Vector[PVarDecl] = {
+        val unsortedDecls: Vector[PVarDecl] = members.collect{ case d: PVarDecl => d }
+        // we require a package to be able to obtain position information
+        val pkgOpt: Option[PPackage] = unsortedDecls.headOption.flatMap(tryEnclosingPackage)
+        // sort declarations by the order in which they appear in the program
+        unsortedDecls.sortBy{ decl =>
+          pkgOpt.get.positions.positions.getStart(decl) match {
+            case Some(pos) => (pos.line, pos.column)
+            case _ => Violation.violation(s"Could not find position information of $decl.")
+          }
+        }
+      }
+      // HACK: without this explicit check, Gobra does not find repeated declarations
```

Comment:
In both methods, we perform extra checks that expect the ids to be well formed. This hack occurs in both functions to guarantee that the checks don't occur in those cases, otherwise they would cause an exception. I can open an issue with improvements that I expect to implement when there is less time pressure and include this as an improvement.